File: typescript/tests/registry.test.ts
import { describe, expect, test } from "vitest";
import { Registry, UnsupportedStepError } from "../src/registry.js";

describe("Registry", () => {
  test("step registers and resolves handlers", () => {
    const reg = new Registry();
    let called = false;
    reg.step("ready", () => {
      called = true;
    });
    expect(reg.has("ready")).toBe(true);
    reg.resolve("ready")({}, {});
    expect(called).toBe(true);
  });

  test("resolve throws UnsupportedStepError on unknown text", () => {
    const reg = new Registry();
    expect(() => reg.resolve("never")).toThrow(UnsupportedStepError);
    try {
      reg.resolve("never");
    } catch (err) {
      expect((err as UnsupportedStepError).text).toBe("never");
    }
  });

  test("step rejects duplicate text", () => {
    const reg = new Registry();
    reg.step("x", () => {});
    expect(() => reg.step("x", () => {})).toThrow(/duplicate step handler/);
  });

  test("has returns false for unknown text", () => {
    const reg = new Registry();
    expect(reg.has("never")).toBe(false);
  });
});

File: go/apskit/registry_test.go
package apskit

import (
	"errors"
	"testing"
)

func TestRegistryStepAndResolve(t *testing.T) {
	r := NewRegistry()
	called := false
	r.Step("ready", func(world World, _ Example) error {
		called = true
		return nil
	})
	fn, err := r.Resolve("ready")
	if err != nil {
		t.Fatalf("resolve: %v", err)
	}
	if err := fn(World{}, Example{}); err != nil {
		t.Fatalf("handler: %v", err)
	}
	if !called {
		t.Fatal("handler was not called")
	}
}

func TestRegistryResolveUnknown(t *testing.T) {
	r := NewRegistry()
	_, err := r.Resolve("never seen")
	var uerr *UnsupportedStepError
	if !errors.As(err, &uerr) {
		t.Fatalf("expected UnsupportedStepError, got %v", err)
	}
	if uerr.Text != "never seen" {
		t.Errorf("unexpected text %q", uerr.Text)
	}
	if got := uerr.Error(); got == "" {
		t.Error("error message should not be empty")
	}
}

func TestRegistryDuplicateStepPanics(t *testing.T) {
	r := NewRegistry()
	r.Step("x", func(World, Example) error { return nil })
	defer func() {
		if recover() == nil {
			t.Fatal("expected panic on duplicate")
		}
	}()
	r.Step("x", func(World, Example) error { return nil })
}

func TestRegistryHas(t *testing.T) {
	r := NewRegistry()
	if r.Has("a") {
		t.Fatal("empty registry should not contain a")
	}
	r.Step("a", func(World, Example) error { return nil })
	if !r.Has("a") {
		t.Fatal("expected to find a")
	}
}
